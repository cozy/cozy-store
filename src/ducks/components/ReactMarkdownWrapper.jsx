import React from 'react'
import ReactMarkdown from 'react-markdown'
import Emoji from 'emoji-js'

const emojiParser = new Emoji()

const parseTitles = props => {
  return React.createElement(
    `h${props.level}`,
    { className: `md-title md-title--h${props.level}` },
    props.children
  )
}

const parseLinks = props => (
  <a
    className="sto-link-text"
    href={props.href}
    target="_blank"
    rel="noopener noreferrer"
  >
    {props.children}
  </a>
)

export const reactMarkdownRendererOptions = {
  link: parseLinks,
  heading: parseTitles
}

export const ReactMarkdownWrapper = ({ source, parseEmoji, className }) => {
  if (!source) return null
  return (
    <ReactMarkdown
      source={parseEmoji ? emojiParser.replace_colons(source) : source}
      renderers={reactMarkdownRendererOptions}
      className={className}
      escapeHtml={false}
    />
  )
}

export default ReactMarkdownWrapper
