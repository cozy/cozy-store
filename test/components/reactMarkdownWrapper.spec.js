import React from 'react'
import { render } from '@testing-library/react'
import '@testing-library/jest-dom'

import ReactMarkdownWrapper, {
  reactMarkdownRendererOptions
} from 'ducks/components/ReactMarkdownWrapper'

describe('ReactMarkdownWrapper component', () => {
  it('should be rendered correctly', () => {
    const { getByText, getByRole } = render(
      <ReactMarkdownWrapper source="Test [linkName]() __strongName__" />
    )
    expect(getByText('Test')).toBeInTheDocument()

    const link = getByRole('link', { name: 'linkName' })
    expect(link).toBeInTheDocument()
    expect(link.getAttribute('class')).toBe('md-link')
    expect(link.getAttribute('href')).toBe('')
    expect(link.getAttribute('rel')).toBe('noopener noreferrer')
    expect(link.getAttribute('target')).toBe('_blank')

    const strong = getByText('strongName')
    expect(strong).toBeInTheDocument()
    expect(strong.tagName).toBe('STRONG')
  })
  it('should handle correctly markdown with emojis', () => {
    const { getByRole, getByText } = render(
      <ReactMarkdownWrapper
        source={'### Heading 3\n\nThis is a `codeName` :tada:'}
        parseEmoji
      />
    )
    expect(
      getByRole('heading', { level: 3, name: 'Heading 3' })
    ).toBeInTheDocument()

    const paragraph = getByText('This is a')
    expect(paragraph).toBeInTheDocument()
    expect(paragraph.tagName).toBe('P')

    const codeName = getByText('codeName')
    expect(codeName).toBeInTheDocument()
    expect(codeName.tagName).toBe('CODE')

    const img = getByRole('img')
    expect(img).toBeInTheDocument()
    expect(img.getAttribute('src')).toBe('/emoji-data/img-apple-64/1f389.png')
  })
  it('should not render if no source provided', () => {
    const { container } = render(<ReactMarkdownWrapper />)
    expect(container).toMatchInlineSnapshot('<div />')
  })
  it('should have correct link renderer options', () => {
    const { getByRole } = render(
      <reactMarkdownRendererOptions.link href="#">
        <div />
      </reactMarkdownRendererOptions.link>
    )

    const link = getByRole('link')
    expect(link).toBeInTheDocument()
    expect(link.getAttribute('class')).toBe('md-link')
    expect(link.getAttribute('href')).toBe('#')
    expect(link.getAttribute('rel')).toBe('noopener noreferrer')
    expect(link.getAttribute('target')).toBe('_blank')
  })
  it('should have correct heading renderer options', () => {
    const { getByRole } = render(
      <reactMarkdownRendererOptions.heading level="3">
        MyHeading
      </reactMarkdownRendererOptions.heading>
    )

    const heading = getByRole('heading', { level: 3, name: 'MyHeading' })
    expect(heading).toBeInTheDocument()
    expect(heading.getAttribute('class')).toBe('md-title md-title--h3')
  })
})
