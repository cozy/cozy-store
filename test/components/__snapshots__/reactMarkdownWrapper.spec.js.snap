// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReactMarkdownWrapper component should be rendered correctly 1`] = `
<ReactMarkdown
  astPlugins={Array []}
  className={undefined}
  escapeHtml={false}
  plugins={Array []}
  rawSourcePos={false}
  renderers={
    Object {
      "heading": [Function],
      "link": [Function],
    }
  }
  skipHtml={false}
  source="Test [link]() __strong__"
  sourcePos={false}
  transformLinkUri={[Function]}
/>
`;

exports[`ReactMarkdownWrapper component should handle correctly markdown with emojis 1`] = `
<ReactMarkdown
  astPlugins={Array []}
  className={undefined}
  escapeHtml={false}
  plugins={Array []}
  rawSourcePos={false}
  renderers={
    Object {
      "heading": [Function],
      "link": [Function],
    }
  }
  skipHtml={false}
  source="### Heading 3

This is a \`test\` <img src=\\"/emoji-data/img-apple-64/1f389.png\\" class=\\"emoji\\" data-codepoints=\\"1f389\\" />"
  sourcePos={false}
  transformLinkUri={[Function]}
/>
`;

exports[`ReactMarkdownWrapper component should have correct heading renderer options 1`] = `
<h3
  className="md-title md-title--h3"
>
  MyHeading
</h3>
`;

exports[`ReactMarkdownWrapper component should have correct link renderer options 1`] = `
<a
  className="sto-link-text"
  href="#"
  rel="noopener noreferrer"
  target="_blank"
>
  <div />
</a>
`;

exports[`ReactMarkdownWrapper component should not render if no source provided 1`] = `null`;
